lib/trie/trie.go

```diff
@@ -167,675 +177,895 @@ func (t *Trie) Entries() map[string][]byte {
+		fullKeyNibbles = append(fullKeyNibbles, prefix...)
+		fullKeyNibbles = append(fullKeyNibbles, branch.Key...)
+		keyLE := string(codec.NibblesToKeyLE(fullKeyNibbles))
+		kv[keyLE] = branch.Value
+	}
+
+	for i, child := range branch.Children {
+		childPrefix := make([]byte, 0, len(prefix)+len(branch.Key)+1)
+		childPrefix = append(childPrefix, prefix...)
+		childPrefix = append(childPrefix, branch.Key...)
+		childPrefix = append(childPrefix, byte(i))
+		entries(child, childPrefix, kv)
+	}
+
 	return kv
 }
 
-// NextKey returns the next key in the trie in lexicographic order. It returns nil if there is no next key
-func (t *Trie) NextKey(key []byte) []byte {
-	k := codec.KeyLEToNibbles(key)
+// NextKey returns the next key in the trie in lexicographic order.
+// It returns nil if no next key is found.
+func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
+	prefix := []byte(nil)
+	key := codec.KeyLEToNibbles(keyLE)
 
-	next := nextKey(t.root, nil, k)
-	if next == nil {
+	nextKey := findNextKey(t.root, prefix, key)
+	if nextKey == nil {
 		return nil
 	}
 
-	return codec.NibblesToKeyLE(next)
+	nextKeyLE = codec.NibblesToKeyLE(nextKey)
+	return nextKeyLE
 }
 
-func nextKey(curr Node, prefix, key []byte) []byte {
-	switch c := curr.(type) {
-	case *node.Branch:
-		fullKey := append(prefix, c.Key...)
-		var cmp int
-		if len(key) < len(fullKey) {
-			if bytes.Compare(key, fullKey[:len(key)]) == 1 { // arg key is greater than full, return nil
-				return nil
-			}
+func findNextKey(parent Node, prefix, searchKey []byte) (nextKey []byte) {
+	if parent == nil {
+		return nil
+	}
 
-			// the key is lexicographically less than the current node key. return first key available
-			cmp = 1
-		} else {
-			// if cmp == 1, then node key is lexicographically greater than the key arg
-			cmp = bytes.Compare(fullKey, key[:len(fullKey)])
-		}
+	if parent.Type() == node.LeafType {
+		parentLeaf := parent.(*node.Leaf)
+		return findNextKeyLeaf(parentLeaf, prefix, searchKey)
+	}
 
-		// if length of key arg is less than branch key,
-		// return key of first child, or key of this branch,
-		// if it's a branch with value.
-		if (cmp == 0 && len(key) == len(fullKey)) || cmp == 1 {
-			if c.Value != nil && bytes.Compare(fullKey, key) > 0 {
-				return fullKey
-			}
+	// Branch
+	parentBranch := parent.(*node.Branch)
+	return findNextKeyBranch(parentBranch, prefix, searchKey)
+}
 
-			for i, child := range c.Children {
-				if child == nil {
-					continue
-				}
+func findNextKeyLeaf(leaf *node.Leaf, prefix, searchKey []byte) (nextKey []byte) {
+	parentLeafKey := leaf.Key
+	fullKey := make([]byte, 0, len(prefix)+len(parentLeafKey))
+	fullKey = append(fullKey, prefix...)
+	fullKey = append(fullKey, parentLeafKey...)
+
+	searchKeyBigger :=
+		(len(searchKey) < len(fullKey) &&
+			bytes.Compare(searchKey, fullKey[:len(searchKey)]) == 1) ||
+			(len(searchKey) >= len(fullKey) &&
+				bytes.Compare(searchKey[:len(fullKey)], fullKey) != -1)
```

Comment:
this really hurts my head and is a lot less readable than before, can you break it up and put comments for the different cases?

Reply:
Totally, done in d8b98631c716c908c8a43a294b98dae20cbf05d7